Archive/Hairer.lean

```diff
@@ -0,0 +1,242 @@
+  zero_mem' :=
+    ⟨(tsupport_eq_empty_iff.mpr rfl).subset.trans (empty_subset _), contDiff_const (c := 0)⟩
+  smul_mem' r f hf :=
+    ⟨(closure_mono <| support_smul_subset_right r f).trans hf.1, contDiff_const.smul hf.2⟩
+
+namespace SmoothSupportedOn
+
+variable {n : ℕ∞} {s : Set E}
+
+instance : FunLike (SmoothSupportedOn 𝕜 E F n s) E (fun _ ↦ F) where
+  coe := Subtype.val
+  coe_injective' := Subtype.coe_injective
+
+@[simp]
+lemma coe_mk (f : E → F) (h) : (⟨f, h⟩ : SmoothSupportedOn 𝕜 E F n s) = f := rfl
+
+lemma tsupport_subset (f : SmoothSupportedOn 𝕜 E F n s) : tsupport f ⊆ s := f.2.1
+
+lemma support_subset (f : SmoothSupportedOn 𝕜 E F n s) :
+    support f ⊆ s := subset_tsupport _ |>.trans (tsupport_subset f)
+
+lemma contDiff (f : SmoothSupportedOn 𝕜 E F n s) :
+    ContDiff 𝕜 n f := f.2.2
+
+theorem continuous (f : SmoothSupportedOn 𝕜 E F n s) : Continuous f :=
+  ContDiff.continuous <| SmoothSupportedOn.contDiff _
+
+lemma hasCompactSupport [ProperSpace E] (f : SmoothSupportedOn 𝕜 E F n (closedBall 0 1)) :
+    HasCompactSupport f :=
+  HasCompactSupport.of_support_subset_isCompact (isCompact_closedBall 0 1) (support_subset f)
+
+end SmoothSupportedOn
+
+end normed
+open SmoothSupportedOn
+
+section missing_polynomial
+open MvPolynomial Submodule
+
+variable {R σ : Type*} [CommSemiring R] (n : ℕ)
+
+lemma restrictTotalDegree_eq_span {n : ℕ} :
+    restrictTotalDegree σ R n =
+    span R ((fun c : σ →₀ ℕ ↦ monomial c (1 : R)) '' {s : σ →₀ ℕ | s.sum (fun _ e ↦ e) ≤ n}) := by
+  ext P; constructor <;> intro h
+  · rw [← P.support_sum_monomial_coeff]
+    refine sum_mem fun c hc ↦ ?_
+    rw [← mul_one (coeff c P), ← smul_eq_mul, ← smul_monomial]
+    rw [mem_restrictTotalDegree] at h
+    exact smul_mem _ _ (subset_span <| mem_image_of_mem _ <| (le_totalDegree hc).trans h)
+  · refine span_le.mpr ?_ h
+    rintro x ⟨c, hc, rfl⟩
+    rw [SetLike.mem_coe, mem_restrictTotalDegree]
+    cases subsingleton_or_nontrivial R
+    · rw [Subsingleton.elim ((fun c ↦ monomial c 1) c) 0, totalDegree_zero]; apply zero_le
+    · rw [totalDegree_monomial _ one_ne_zero]; exact hc
+
+/- Can be obtained by combining `LinearMap.proj` and `MvPolynomial.evalₗ`. -/
+/-- The evaluation of a multivariate polynomial at a point, as a linear map. -/
+def evalAtₗ {R σ : Type*} [CommSemiring R] (x : σ → R) : MvPolynomial σ R →ₗ[R] R where
+  toFun P := eval x P
+  map_add' := by simp
+  map_smul' := by simp
+
+/- what we actually want -/
+lemma analyticOn_eval {R σ : Type*} [Fintype σ] [IsROrC R] (P : MvPolynomial σ R) :
+    AnalyticOn R (fun x ↦ eval (EuclideanSpace.equiv σ R x) P) univ := fun x _ ↦ by
+  apply P.induction_on (fun r ↦ ?_) (fun p q hp hq ↦ ?_) fun p i hp ↦ ?_ -- refine doesn't work
+  · simp_rw [eval_C]; exact analyticAt_const
+  · simp_rw [eval_add]; exact hp.add hq
+  · simp_rw [eval_mul, eval_X]; exact hp.mul ((ContinuousLinearMap.proj (R := R) i).analyticAt x)
```

Comment:
This should be an easy PR but needs a new file because there's nothing that imports both MvPolynomial and AnalyticOn. I think the correct generality is that you can eval on a family of analytic functions `σ → E → F` for normed spaces `E` and `F` over a NontriviallyNormedField R. If `E` is finite dimensional Hausdorff, any linear `E → F` is continuous, so you can state the result with a linear map `E →ₗ[R] σ → F`.

```suggestion
lemma analyticOn_eval {R σ : Type*} [Fintype σ] [IsROrC R] (P : MvPolynomial σ R) :
    AnalyticOn R (fun x ↦ eval (EuclideanSpace.equiv σ R x) P) univ := fun x _ ↦ by
  apply P.induction_on (fun r ↦ ?_) (fun p q hp hq ↦ ?_) fun p i hp ↦ ?_ -- refine doesn't work
  · simp_rw [eval_C]; exact analyticAt_const
  · simp_rw [eval_add]; exact hp.add hq
  · simp_rw [eval_mul, eval_X]; exact hp.mul ((ContinuousLinearMap.proj (R := R) i).analyticAt x)
```

Reply:
This is now in another file (could be PR'd separately if necessary)